[FLINK-9116] Introduce getAll and removeAll for MapState

klion26 opened the pull request:
## What is the purpose of the change

Introduce getAll and removeAll for MapState


## Verifying this change

This change added unit tests 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

StefanRRichter commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/mock/MockInternalMapState.java:
```diff
@@ -46,6 +47,18 @@ public UV get(UK key) {
+
+		Map<UK, UV> userMaps = getInternal();
+
+		Map<UK, UV> maps = new HashMap<>();
```
Can be initialized with capacity `keys.size()`

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java:
```diff
@@ -91,6 +93,22 @@ public UV get(UK userKey) {
+			return Collections.emptyMap();
+		}
+
+		Map<UK, UV> maps = new HashMap<>();
```
Can also be initialized with capacity `keys.size()`

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java:
```diff
@@ -91,6 +93,22 @@ public UV get(UK userKey) {
+	@Override
+	public Map<UK, UV> getAll(Collection<UK> keys) {
+
+		Map<UK, UV> userMap = stateTable.get(currentNamespace);
```
I feel like by now we could really introduce a helper method `getUserMap() {stateTable.get(currentNamespace);}`.

StefanRRichter commented on flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/client/state/ImmutableMapState.java:
```diff
@@ -52,6 +53,16 @@ public V get(K key) {
+	@Override
+	public Map<K, V> getAll(Collection<K> keys) {
+
+		Map<K, V> maps = new HashMap<>();
```
Again init with capacity.

StefanRRichter commented on flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/utils/TestSharedBuffer.java:
```diff
@@ -151,6 +152,16 @@ public UV get(UK key) throws Exception {
 
+				@Override
+				public Map<UK, UV> getAll(Collection<UK> keys) throws Exception {
+					Map<UK, UV> maps = new HashMap<>();
```
Same here.

StefanRRichter commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java:
```diff
@@ -129,6 +132,36 @@ public UV get(UK userKey) throws IOException, RocksDBException {
+			keyBytesMap.put(key, keyByte);
+		}
+
+		Map<byte[], byte[]> result = backend.db.multiGet(columnFamilyHandles, keyBytesList);
```
I think we should favor iterating individual keys over adding them all to a list & multiGet. It seems more more memory efficient.

https://github.com/facebook/rocksdb/wiki/RocksDB-FAQ

> Q: If I want to retrieve 10 keys from rocksdb, is it better to batch them and use MultiGet() versus issuing 10 individual Get() calls?
> 
> A: The performance is similar. MultiGet() reads from the same consistent view, but it is not faster.

klion26 commented:
Hi, @StefanRRichter I have addressed your comments. Please have another review.

In my opinion, add getAll/removeAll interface is convenience for these who want to get all values for a collection of keys, or remove a collection of keys from the state.

klion26 commented:
Hi, @StefanRRichter If I use a field to store the user map in Heap backend, I will override the `AbstractHeapState#clear` to update the field, so I'll change `public final void clear()` in AbstractHeapState to `public void clear()` , and change the `getUserMap` in Heap backend such as `if (userMaps == null) {
			userMaps = stateTable.get(currentNamespace);
		} else if (stateTable.keyContext.getCurrentKey() != currentKey) {
			userMaps = stateTable.get(currentNamespace);
			currentKey = stateTable.keyContext.getCurrentKey();
		}
		return userMaps;` is this ok?
